packages/builder/src/interface.ts

```diff
@@ -0,0 +1,417 @@
+  <T extends JSONSchema = JSONSchema, R extends JSONSchema = JSONSchema>(
+    argumentSchema: T,
+    resultSchema: R,
+    implementation: (input: Schema<T>) => Schema<R>,
+  ): ModuleFactory<Schema<T>, Schema<R>>;
+
+  <T, R>(
+    implementation: (input: T) => R,
+  ): ModuleFactory<T, R>;
+
+  <T>(
+    implementation: (input: T) => any,
+  ): ModuleFactory<T, ReturnType<typeof implementation>>;
+
+  <T extends (...args: any[]) => any>(
+    implementation: T,
+  ): ModuleFactory<Parameters<T>[0], ReturnType<T>>;
+};
+
+export type HandlerFunction = {
+  <E extends JSONSchema = JSONSchema, T extends JSONSchema = JSONSchema>(
+    eventSchema: E,
+    stateSchema: T,
+    handler: (event: Schema<E>, props: Schema<T>) => any,
+  ): ModuleFactory<Schema<T>, Schema<E>>;
+
+  <E, T>(
+    eventSchema: JSONSchema,
+    stateSchema: JSONSchema,
+    handler: (event: E, props: T) => any,
+  ): ModuleFactory<T, E>;
+
+  <E, T>(
+    handler: (event: E, props: T) => any,
+  ): ModuleFactory<T, E>;
+};
+
+export type DeriveFunction = <In, Out>(
+  input: Opaque<In>,
+  f: (input: In) => Out | Promise<Out>,
+) => OpaqueRef<Out>;
+
+export type ComputeFunction = <T>(fn: () => T) => OpaqueRef<T>;
+
+export type RenderFunction = <T>(fn: () => T) => OpaqueRef<T>;
+
+export type StrFunction = (
+  strings: TemplateStringsArray,
+  ...values: any[]
+) => OpaqueRef<string>;
+
+export type IfElseFunction = <T = any, U = any, V = any>(
+  condition: Opaque<T>,
+  ifTrue: Opaque<U>,
+  ifFalse: Opaque<V>,
+) => OpaqueRef<U | V>;
+
+export type LLMFunction = <T = string>(
+  params: Opaque<BuiltInLLMParams>,
+) => OpaqueRef<BuiltInLLMState<T>>;
+
+export type FetchDataFunction = <T>(
+  params: Opaque<{
+    url: string;
+    mode?: "json" | "text";
+    options?: RequestInit;
+    result?: T;
+  }>,
+) => Opaque<{ pending: boolean; result: T; error: any }>;
+
+export type StreamDataFunction = <T>(
+  params: Opaque<{
+    url: string;
+    options?: RequestInit;
+    result?: T;
+  }>,
+) => Opaque<{ pending: boolean; result: T; error: any }>;
+
+export type CompileAndRunFunction = <T = any, S = any>(
+  params: Opaque<BuiltInCompileAndRunParams<T>>,
+) => OpaqueRef<BuiltInCompileAndRunState<S>>;
+
+export type NavigateToFunction = (cell: OpaqueRef<any>) => OpaqueRef<string>;
+
+export type CreateNodeFactoryFunction = <T = any, R = any>(
+  moduleSpec: Module,
+) => ModuleFactory<T, R>;
+
+export type CreateCellFunction = {
+  <T>(
+    schema?: JSONSchema,
+    name?: string,
+    value?: T,
+  ): Cell<T>;
+
+  <S extends JSONSchema = JSONSchema>(
+    schema: S,
+    name?: string,
+    value?: Schema<S>,
+  ): Cell<Schema<S>>;
+};
+
+// Re-export opaque ref creators
+export type CellFunction = <T>(value?: T, schema?: JSONSchema) => OpaqueRef<T>;
+export type StreamFunction = <T>(initial?: T) => OpaqueRef<T>;
+export type ByRefFunction = <T, R>(ref: string) => ModuleFactory<T, R>;
+
+// Builder functions interface
+export interface BuilderFunctions {
```

Comment:
If this isn't used directly by recipes, consider moving it outside of interface

Reply:
Done!